src/cpp/shared/town/town.cpp

```diff
@@ -1434,70 +1436,57 @@ void townManager::SetupCastle(heroWindow *window, int a3) {
   }
 }
 
-int __fastcall CanBuild(town *twn, int building) {
-  
-  int result; // eax@2
-  
+int __fastcall CanBuild(town *twn, int building) {    
   if(BitTest((const LONG*)gpGame->builtToday, twn->idx)) {
-    result = 0;
-  } else {
-    int v6 = building;
-    if(v6 == 6 || twn->buildingsBuiltFlags & 0x40) {
-      if(xIsExpansionMap || v6 != 2 || twn->factionID != 5) {
-        if(v6 == 3) {
-          if(twn->CanBuildDock())
-            result = 1;
-          else
-            result = 0;
-        } else {
-          if(v6 || twn->mageGuildLevel < 5) {
-            if(v6 != 5 && v6 != 14 && v6 != 16 && v6 != 17 && v6 != 18 && v6 != 31) {
-              if(v6 >= 19 && v6 <= 30) {
-                unsigned int byte3 = *((unsigned char*)&(twn->buildingsBuiltFlags)+3);                  
-                if(v6 == 20 && byte3 & 2
-                  || v6 == 21 && byte3 & 4
-                  || v6 == 22 && byte3 & 8
-                  || v6 == 23 && byte3 & 0x10
-                  || v6 == 24 && (byte3 & 0x20 || byte3 & 0x40)
-                  || v6 == 29 && byte3 & 0x40) {
-                  result = 0;
-                } else {
-                  unsigned int v8 = twn->buildingsBuiltFlags;
-                  unsigned int byte3 = *((unsigned char*)&(v8)+3);    
-                  if(byte3 & 2)
-                    v8 |= 0x100000u;
-                  if(byte3 & 4)
-                    v8 |= 0x200000u;
-                  if(byte3 & 8)
-                    v8 |= 0x400000u;
-                  if(byte3 & 0x10)
-                    v8 |= 0x800000u;
-                  if(byte3 & 0x40)
-                    v8 |= 0x20000000u;
-                  if(byte3 & 0x20)
-                    v8 |= 0x1000000u;
-                  int v5 = gHierarchyMask[twn->factionID][v6 - 19]; // - 19
-                  if((v8 & v5) == v5)
-                    result = twn->factionID != 5 || v6 != 28 || twn->mageGuildLevel > 1;
-                  else
-                    result = 0;
-                }
-              } else {
-                result = 1;
-              }
-            } else {
-              result = 0;
+    return 0;
+  }
+
+  unsigned int builtFlags = twn->buildingsBuiltFlags;
+  if(building == BUILDING_CASTLE || builtFlags & 0x40) {
+    if(xIsExpansionMap || building != BUILDING_TAVERN || twn->factionID != FACTION_NECROMANCER) {
+      if(building == BUILDING_DOCK) {
+        if(twn->CanBuildDock())
+          return 1;
+        return 0;
+      }
+
+      if(building || twn->mageGuildLevel < 5) {
+        if(building != BUILDING_TENT && building != BUILDING_BOAT && building != BUILDING_EXT_0 && building != BUILDING_EXT_1 && building != BUILDING_EXT_2 && building != BUILDING_EXT_3) {
+          if(building >= BUILDING_DWELLING_1 && building <= BUILDING_UPGRADE_5B) {
+            unsigned int byte3 = *((unsigned char*)&(builtFlags)+3);
+            if(building == BUILDING_DWELLING_2 && byte3 & 2
```

Comment:
Same kind of bit test refactoring I mentioned.  This line, for example, is: building == BUILDING_DWELLING_2 && (builtFlags & BUILDING_UPGRADE_1).